services/auth/httpsign.go

```diff
@@ -0,0 +1,183 @@
+	"code.gitea.io/gitea/modules/setting"
+	"code.gitea.io/gitea/modules/web/middleware"
+
+	"github.com/go-fed/httpsig"
+	"golang.org/x/crypto/ssh"
+)
+
+// Ensure the struct implements the interface.
+var (
+	_ Method = &HTTPSign{}
+	_ Named  = &HTTPSign{}
+)
+
+// HTTPSign implements the Auth interface and authenticates requests (API requests
+// only) by looking for http signature data in the "Signature" header.
+// more information can be found on https://github.com/go-fed/httpsig
+type HTTPSign struct{}
+
+// Name represents the name of auth method
+func (h *HTTPSign) Name() string {
+	return "httpsign"
+}
+
+// Verify extracts and validates HTTPsign from the Signature header of the request and returns
+// the corresponding user object on successful validation.
+// Returns nil if header is empty or validation fails.
+func (h *HTTPSign) Verify(req *http.Request, w http.ResponseWriter, store DataStore, sess SessionStore) *user_model.User {
+	// HTTPSign authentication should only fire on API
+	if !middleware.IsAPIPath(req) {
+		return nil
+	}
+
+	sigHead := req.Header.Get("Signature")
+	if len(sigHead) == 0 {
+		return nil
+	}
+
+	if len(setting.SSH.TrustedUserCAKeys) == 0 {
+		return nil
+	}
+
+	validpk, err := VerifyCert(req)
+	if err != nil {
+		log.Warn("Failed authentication attempt from %s", req.RemoteAddr)
+		log.Error("VerifyCert failed: %v", err)
+		return nil
+	}
+
+	u, err := user_model.GetUserByID(validpk.OwnerID)
+	if err != nil {
+		log.Error("GetUserByID:  %v", err)
+		return nil
+	}
+
+	store.GetData()["IsApiToken"] = true
+
+	log.Trace("HTTP Sign: Logged in user %-v", u)
+
+	return u
+}
+
+// VerifyCert verifies the validity of the ssh certificate and returns the publickey of the signer
+// We verify that the certificate is signed with the correct CA
+// We verify that the http request is signed with the private key (of the public key mentioned in the certificate)
+func VerifyCert(r *http.Request) (*asymkey_model.PublicKey, error) {
+	var validpk *asymkey_model.PublicKey
+
+	// Get our certificate from the header
+	bcert, err := base64.RawStdEncoding.DecodeString(r.Header.Get("x-ssh-certificate"))
+	if err != nil {
+		return validpk, err
+	}
+
+	pk, err := ssh.ParsePublicKey(bcert)
+	if err != nil {
+		return validpk, err
+	}
+
+	// Check if it's really a ssh certificate
+	cert, ok := pk.(*ssh.Certificate)
+	if !ok {
+		return validpk, fmt.Errorf("no certificate found")
+	}
+
+	for _, principal := range cert.ValidPrincipals {
+		validpk, err = asymkey_model.SearchPublicKeyByContentExact(r.Context(), principal)
+		if err != nil {
+			if asymkey_model.IsErrKeyNotExist(err) {
+				continue
+			}
+			log.Error("SearchPublicKeyByContentExact: %v", err)
+			return validpk, err
+		}
+
+		c := &ssh.CertChecker{
+			IsUserAuthority: func(auth ssh.PublicKey) bool {
+				for _, k := range setting.SSH.TrustedUserCAKeysParsed {
+					if bytes.Equal(auth.Marshal(), k.Marshal()) {
+						return true
+					}
+				}
+
+				return false
+			},
```

Comment:
Marshal `auth` once and once only. Do not repeatedly run `auth.Marshal()`.

(I've just noticed that modules/ssh/ssh.go does the same thing - can we change it there too?)

```suggestion
			IsUserAuthority: func(auth ssh.PublicKey) bool {
			  marshaled := auth.Marshal()
				for _, k := range setting.SSH.TrustedUserCAKeysParsed {
					if bytes.Equal(marshaled, k.Marshal()) {
						return true
					}
				}

				return false
			},
```

Reply:
fair enough, I can make a separate PR for the fix to the ssh one ?

Reply:
Just do it in here I think.

Reply:
ok